src/mlir/generator.cc

```diff
@@ -621,6 +652,75 @@ namespace mlir::verona
     return ReturnValue();
   }
 
+  llvm::Expected<ReturnValue> Generator::parseForLoop(const ::ast::Ast& ast)
+  {
+    assert(isFor(ast) && "Bad node");
+
+    // For loops are of the shape (item in list), which need to initialise
+    // the item, take the next from the list and exit if there is none.
+    // All of that is within the scope of the loop, so we need to create a
+    // scope now, to drop them from future ops.
+    SymbolScopeT var_scope{symbolTable};
+
+    // First, we identify the iterator. Nested loops will have their own
+    // iterators, of the same name, but within their own lexical blocks.
+    auto seqNode = getLoopSeq(ast).lock();
+    auto list = parseNode(seqNode);
+    if (auto err = list.takeError())
+      return std::move(err);
+    declareVariable("$iter", list->get());
+    llvm::SmallVector<mlir::Value, 1> iter{symbolTable.lookup("$iter")};
+
+    // Create the head basic-block, which will check the condition
+    // and dispatch the loop to the body block or exit.
+    auto region = builder.getInsertionBlock()->getParent();
+    mlir::ValueRange empty{};
+    auto headBB = addBlock(region);
+    auto bodyBB = addBlock(region);
+    auto exitBB = addBlock(region);
+    builder.create<mlir::BranchOp>(getLocation(ast), headBB, empty);
+
+    // First node is a check on the lists `has_value` boolean return.
+    builder.setInsertionPointToEnd(headBB);
+    auto condLoc = getLocation(seqNode);
+    auto has_value = functionTable.lookup("has_value");
```

Comment:
Mildly worried about hard coded strings here `"has_value"`

Reply:
Me too. But I don't have a better idea for the moment. This is part of the language ABI, so it cannot be different (for whatever name we choose), but this knowledge should not be here, and should be in an ABI handler.

This ABI handler would not only know what the function names and signatures be, it would also declare them beforehand and return the signatures when requested by the lowering mechanism.

I haven't thought too much about such a mechanism and would like a larger discussion around this, as we'll need to add many more things around language ABI.

See #259.

For now, I'll define them as constants in the Generator class (like the types and other ABI stuff). We can move everything out later in a different PR.